Updated to sanitize asset filename prior to uploading.

jtpetty opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->
Fixes #7024 

<img width="926" alt="image" src="https://user-images.githubusercontent.com/6889008/226353598-04855bfc-4be2-4c29-b931-87b57810edce.png">

heaths committed:
Retry fetching repo from template

Fixes #7055

srgk26 committed:
Added apt update command to linux install docs.

wingkwong committed:
feat: add gistRenameCmd to gist.go

wingkwong committed:
feat: add gist rename logic

wingkwong committed:
refactor: remove unused code

wingkwong committed:
feat: TestNewCmdRename for gist rename

wingkwong committed:
feat: add FlagErrorf for missing flag cases

wingkwong committed:
fix: revise test cases in TestNewCmdRename in rename_test.go

jtpetty committed:
Updated to sanitize asset filename prior to uploading. Fixes #7024

jtpetty committed:
Merge branch 'trunk' into jtpetty/issue-7024

mislav commented on pkg/cmd/release/upload/upload.go:
```diff
@@ -122,3 +125,28 @@ func uploadRun(opts *UploadOptions) error {
+
+func sanitizeFileName(name string) string {
+
+	value := regexp.MustCompile("[[:^ascii:]]").ReplaceAllLiteralString(name, "")
```
This is not exactly what the platform does, since the platform transliterates non-ASCII Unicode characters to their Latin equivalents (e.g. converts `ć` in my name to `c`). We could pull in a transliteration library according to https://github.com/cli/cli/issues/7142#issuecomment-1465907659, or we could opt to do nothing in this case, since I'm guessing it's a rare circumstance that someone will use `--clobber` with file names that contains non-ASCII characters, but I don't think that replacing non-ASCII characters with nothing is useful here.

What do you think would be a good approach?

mislav replied:
BTW I just found transliteration rules that Rails uses; they look simple enough: https://github.com/ruby-i18n/i18n/blob/b4cd00d73234e369d713eb2607c729765bb3ec06/lib/i18n/backend/transliterator.rb#L44-L71

mislav replied:
Update: a transliteration effort in progress in another PR by a contributor: https://github.com/cli/cli/pull/7205#discussion_r1145429247

mislav replied:
@jtpetty Update: the function to remove diacritics from a string is now in the trunk branch: `text.RemoveDiacritics()`

mislav commented on pkg/cmd/release/upload/upload.go:
```diff
@@ -122,3 +125,28 @@ func uploadRun(opts *UploadOptions) error {
 	return nil
 }
+
+func sanitizeFileName(name string) string {
```
Nit: this function could benefit from a comment that explains that this tries to mimic on the client what the platform does for uploaded assets

jtpetty replied:
👍

mislav commented on pkg/cmd/release/upload/upload.go:
```diff
@@ -122,3 +125,28 @@ func uploadRun(opts *UploadOptions) error {
+	value = regexp.MustCompile(`\.{2,}`).ReplaceAllLiteralString(value, ".")
+
+	// Remove leading/trailing separator.
+	value = regexp.MustCompile(`(?i)\A\.|\.\z`).ReplaceAllLiteralString(value, "")
```
Protip:
```suggestion
	value = strings.Trim(value, ".")
```

jtpetty replied:
👍

mislav commented on pkg/cmd/release/upload/upload.go:
```diff
@@ -93,9 +94,11 @@ func uploadRun(opts *UploadOptions) error {
-			if ea.Name == a.Name {
+			if ea.Name == sanitizedFileName {
 				a.ExistingURL = ea.APIURL
+				a.Name = ea.Name
```
Is it necessary to override the `Name` field here? To my knowledge, only ExistingURL field is used when deleting assets. For other code that might operate on `opts.Assets`, it might be good if `a.Name` remained unchanged.

jtpetty replied:
good call, for some reason when I was testing this earlier, I thought it needed this, but I removed it and it seems to work correctly now

mislav reviewed (changes requested):
Thanks! This looks good, but I have some style suggestions that should be easily addressed

heaths committed:
Support ext install --force (#7173)

Resolves #7096

rsteube committed:
variable: fix listing and deleting variables in an environment (#7207)

Co-authored-by: Mislav Marohnić <mislav@github.com>

jtpetty committed:
Addressed some comments from review. Still need to fix non-ascii substitution

wingkwong committed:
feat: gist rename test

dependabot[bot] committed:
Bump github.com/mattn/go-isatty from 0.0.17 to 0.0.18 (#7217)

Bumps [github.com/mattn/go-isatty](https://github.com/mattn/go-isatty) from 0.0.17 to 0.0.18.
- [Release notes](https://github.com/mattn/go-isatty/releases)
- [Commits](https://github.com/mattn/go-isatty/compare/v0.0.17...v0.0.18)

---
updated-dependencies:
- dependency-name: github.com/mattn/go-isatty
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

samcoe committed:
Properly handle closing files that have been writen to (#7199)

wingkwong committed:
refactor: change from [] to {}

wingkwong committed:
refactor: use ExactArgs instead of MaximumNArgs

wingkwong committed:
refactor: remove unnecessary validations

wingkwong committed:
refactor: revise errMsg in rename_test.go

wingkwong committed:
feat: add attempt flag to run view

wingkwong committed:
feat: pass 0 to fetch latest attempt

wingkwong committed:
feat: update GetRun path based on attempt

wingkwong committed:
feat: pass attempt to RenderRunHeader

wingkwong committed:
feat: add attemptLabel to RenderRunHeader

wingkwong committed:
feat: pass latest attempt to RenderRunHeader for watch

wingkwong committed:
feat: append attempt number in returned run url

wingkwong committed:
refactor: don't show latest attempt

wingkwong committed:
feat: include attempt in getRunLog

wingkwong committed:
feat: add tests for attempt flag

wingkwong committed:
docs: revise GH_CONFIG_DIR content in help_topic.go (#7230)

benjlevesque committed:
Add a `--fail-fast` option to `pr checks --watch` (#7203)

guerinoni committed:
Add `--template` flag for issue create and pr create (#7185)

Signed-off-by: Federico Guerinoni <guerinoni.federico@gmail.com>
Co-authored-by: Alessio Cosenza <cosenzaalessio17@gmail.com>

wingkwong committed:
fix: change MaximumNArgs from 2 to 1 in run view.go

wingkwong committed:
refactor: set the default case first

wingkwong committed:
refactor: use net/url JoinPath

wingkwong committed:
fix: throw error for non-existing org / repo with non-zero status in repo list (#7240)

macmacbr committed:
repo sync: clearer error message with actionable hint (#7110)

Co-authored-by: Mislav Marohnić <mislav@github.com>

mislav committed:
repo list: add test for invalid owner error

Commit:
improve test resiliency

heaths committed:
Specify branch to force exit code

heaths committed:
Resolve lint error

bchadwic committed:
Autocomplete branch flags (#6031)

Co-authored-by: Mislav Marohnić <mislav@github.com>

SonicGDX committed:
Fix typo in README.md

Signed-off-by: SonicGDX<114670430+SonicGDX@users.noreply.github.com>

vaindil committed:
check SSH key existence before uploading

joshkraft committed:
feat: add gh org list

alex-petrov-vt committed:
feat: allow filtering workflow runs by statuses

mislav committed:
Improve Amazon Linux install instructions

phillippbertram committed:
feat(run): add new delete command

phillippbertram committed:
cleanup

phillippbertram committed:
lint fixes

vaindil committed:
don't parse key bytes twice

vaindil committed:
move SSH key errors out of upload function

kkocha committed:
docs: add commit SHA arg to gh browse help (#7267)

Co-authored-by: Mislav Marohnić <mislav@github.com>

kousikmitra committed:
Add option to add ssh singing key

kkocha committed:
Additional help doc and example for auth setup-git (#7243)

kkocha committed:
fix: make number arg, commit arg, and flags mutually exclusive (#7268)

* make number, commit, and flags mutually exclusive

* break condition into arg own check

Commit:
add --insecure-storage and make secure storage the default

Commit:
fix apparent typo

Commit:
extend secure storage default to auth refresh

kousikmitra committed:
Enhance ssh-key list to show signing ssh keys also

joshkraft committed:
add "showing x of y organizations" header

cmbrose committed:
Fix invoker immediately closing

kkocha committed:
Add `-R` as shorthand of `--repo` flag for search commands (#7274)

Xerkus committed:
Generate completion scripts for bundling in rpm and deb on release

Xerkus committed:
Move completions generation from goreleaser to Makefile

kousikmitra committed:
Enable owner flag to take multiple values for searching

doaortu committed:
feat: add web flag for codespace list & create subcommand (#7288)

* feat: add web flag to cs list subcommand

web flag only works with repo flag, because,
currently there only param for listing with repo_id

* feat: add web flag to cs crate subcommand

web flag used for creating codespace through web UI instead of terminal.
web flag cannot be used with display-name, idle-timeout,
or retention retention-period
because there's no option for that in the Web UI

* refactor: extract mutual excusive logic to PreRunE

- changed web flag mutual exclusive logic, using cmdutil
- extract that logic to PreRunE clause in createCmd
- move web flag up to make it close to PreRunE clause (for clarity)
- add new param to newCreateCmd fn to facilitate test logic
- apply new newCreateCmd fn to root.go

* fix: clarify flag desc and error message

- remove 'yet' from error messages that can cause misunderstanding
- clarify list web flag can only be used with repo flag

* fix: skip machine check when we use web flag ...
(..and no machine flag provided)

+ add test for this new case
+ adjust related test cases for this new change.

* refactor: move flag check logic to PreRunE

why: err on PreRunE or RunE will also print help if error happened

+ move web, repo, org, user mutual exclusive logic to PreRunE clause
+ move repo, org, user mutual exclusive logic to PreRunE
+ move limit check flag to PreRunE
+ modify newListCmd fn to facilitate test logic
+ apply new newListCmd to root.go
+ add test cases to check PreRunE clause
- remove mutual exclusive test cases from Test_AppList

* refactor: remove the opts equality checks

* fix: mutually exclusive misfires
because of wrong logic

+ refine test case too

* cleanup:removing useWeb check in fn getMachineName

because it's no longer needed

+ remove redundant test-case

* refactor: remove redundant ifs

* refactor: clarify test name

* re-clarify web flag desc in list.go

* refactor: break long lines, use more idiomatic err

* add test case for nonexistent/wrong machine

FalseDev committed:
Fix typo in `cs stop` command: `Stoppping` -> `Stopping` (#7318)

samcoe committed:
Update go-gh to v2 (#7299)

* Update go-gh

* Update code for go-gh v2

tuananhlai committed:
fix: normalize hostname for auth login (#6999)

Co-authored-by: Mislav Marohnić <mislav@github.com>

dependabot[bot] committed:
build(deps): bump github.com/cenkalti/backoff/v4 from 4.2.0 to 4.2.1

Bumps [github.com/cenkalti/backoff/v4](https://github.com/cenkalti/backoff) from 4.2.0 to 4.2.1.
- [Release notes](https://github.com/cenkalti/backoff/releases)
- [Commits](https://github.com/cenkalti/backoff/compare/v4.2.0...v4.2.1)

---
updated-dependencies:
- dependency-name: github.com/cenkalti/backoff/v4
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

jkeech committed:
Clarify how SSH keys are selected (#7325)

Fixes: https://github.com/orgs/community/discussions/52556

Update the help text for `gh codespace ssh` to provide more details about how SSH keys are selected.

mislav committed:
Initialize deployment.yml workflow file

cmbrose committed:
Fix `gh cs ports` requiring `sudo` for privileged port numbers (#7326)

goggle committed:
Correct some typos

benjlevesque committed:
Diacritics substitution in prompt (#7205)

cawfeecake committed:
add `--event` flag to `gh run ls` to filter runs by their trigger

cawfeecake committed:
add `--created` flag to `gh run ls` to filter listed runs by when they were created (#7352)

heaths committed:
Edit multiple issues, PRs in parallel (#7259)

Allows multiple issues or PRs to be edited in parallel, and querying for shared fields once to reduce network requests.

Co-authored-by: Sam Coe <samcoe@users.noreply.github.com>

yeikel committed:
feat: add visibility field (#7337)

SauravMaheshkar committed:
gh: move `CODEOWNERS` inside the `.github/` dir (#7366)

mjpieters committed:
Pretty-print JSON results of jq filtering (#7236)

When connected to a TTY, tell the jq formatter to indent the output, and
enable colorization of the output if the terminal supports it.

Co-authored-by: Mislav Marohnić <mislav@github.com>

mislav committed:
New deployment workflow

Add ability to trigger deployments without having to push a git tag

Xerkus committed:
Fix Makefile compatibility with macOS

`install` command on macOS does not support all flags supported by coreutils version.
Create directories as separate `install -d` call.

Signed-off-by: Aleksei Khudiakov <aleksey@xerkus.pro>

Biswa96 committed:
Make: Fix target name for Windows platform (#7370)

This fixes the taget name by adding .exe extension for Windows platform.
Otherwise, the following error is shown with `make bin/gh.exe' command.

    make: *** No rule to make target 'script/build', needed by 'bin/gh.exe'.  Stop.

kousikmitra committed:
Add capability to filter codespaces by repo owner (#7347)

* Add capability to filter codespaces by repo owner

* Replace Flags with PersistentFlags

* Reword flag description

* Update test seed

* Add mutual exclusion

---------

Co-authored-by: Caleb Brose <5447118+cmbrose@users.noreply.github.com>

cmbrose committed:
Fix shellcheck bugs in generate.sh (#7392)

kevhlee committed:
Add `alias import` command (#7118)

vwkd committed:
fix empty default description for local repository (#7383)

mjpieters committed:
Add PR auto-merge status info to PR queries (#7384)

joshkraft committed:
Add progress labels for `gh release download` (#7380)

mjpieters committed:
Include auto-merge information in gh pr status (#7386)

PRs that have auto-merge enabled are marked with a green "✓ Auto-merge enabled" label.

mjpieters committed:
Include auto-merge information in gh pr view (#7385)

kousikmitra committed:
Print message after codespace deletion (#7353)

mislav committed:
Fix release guard for deployment workflow

Co-authored-by: Sam Coe <samcoe@users.noreply.github.com>

mislav committed:
Fix label assets

williammartin committed:
Introduce helpTopics type and reduce duplication across commands (#7414)

Jackenmen committed:
Allow `gh repo set-default --view` without repo argument (#7441)

ffalor committed:
respect GH_REPO env variable in `pr create`

rick-kilgore committed:
survey: fix color contrast of default values in prompts (#7354)

goldstar611 committed:
Mention setting GH_ENTERPRISE_TOKEN

When GH_HOST is set

samcoe committed:
Do not fall back to legacy template if template selector returns nil (#7444)

williammartin committed:
Add some further test coverage around PR ChecksStatus

williammartin committed:
Add some comments to PR ChecksStatus

williammartin committed:
Avoid using named return in PR ChecksStatus

See: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_named_return_values

mislav committed:
Fix windows crash by bumping wincred

williammartin committed:
Remove old, unused detector fields (#7458)

dependabot[bot] committed:
build(deps): bump github.com/stretchr/testify from 1.8.1 to 1.8.2

Bumps [github.com/stretchr/testify](https://github.com/stretchr/testify) from 1.8.1 to 1.8.2.
- [Release notes](https://github.com/stretchr/testify/releases)
- [Commits](https://github.com/stretchr/testify/compare/v1.8.1...v1.8.2)

---
updated-dependencies:
- dependency-name: github.com/stretchr/testify
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump github.com/stretchr/testify from 1.8.2 to 1.8.3

Bumps [github.com/stretchr/testify](https://github.com/stretchr/testify) from 1.8.2 to 1.8.3.
- [Release notes](https://github.com/stretchr/testify/releases)
- [Commits](https://github.com/stretchr/testify/compare/v1.8.2...v1.8.3)

---
updated-dependencies:
- dependency-name: github.com/stretchr/testify
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

williammartin committed:
Update triage.md with severity descriptions

torgeir committed:
Clarify docs regarding what is listed when running gh workflow list

Prevent confusion that `--all` actually list all workflows, `--limit` is
what actually needs to be adjusted to list all workflows, if there are
many.

dependabot[bot] committed:
build(deps): bump github.com/mattn/go-isatty from 0.0.18 to 0.0.19

Bumps [github.com/mattn/go-isatty](https://github.com/mattn/go-isatty) from 0.0.18 to 0.0.19.
- [Commits](https://github.com/mattn/go-isatty/compare/v0.0.18...v0.0.19)

---
updated-dependencies:
- dependency-name: github.com/mattn/go-isatty
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

jtpetty committed:
Use text.RemoveDiacritics()

jtpetty commented:
Had a bad rebase, so created https://github.com/cli/cli/pull/7477 to carry this forward. Closing this PR.
